base/compiler/inferenceresult.jl

```diff
@@ -4,90 +4,107 @@ const EMPTY_VECTOR = Vector{Any}()
 
 mutable struct InferenceResult
     linfo::MethodInstance
-    args::Vector{Any}
-    vargs::Vector{Any} # Memoize vararg type info w/Consts here when calling get_argtypes
-                       # on the InferenceResult, so that the optimizer can use this info
-                       # later during inlining.
+    argtypes::Vector{Any}
     result # ::Type, or InferenceState if WIP
     src #::Union{CodeInfo, OptimizationState, Nothing} # if inferred copy is available
-    function InferenceResult(linfo::MethodInstance)
+    function InferenceResult(linfo::MethodInstance, given_argtypes = nothing)
         if isdefined(linfo, :inferred_const)
             result = Const(linfo.inferred_const)
         else
             result = linfo.rettype
         end
-        return new(linfo, EMPTY_VECTOR, Any[], result, nothing)
+        return new(linfo, matching_cache_argtypes(linfo, given_argtypes), result, nothing)
     end
 end
 
-function get_argtypes(result::InferenceResult)
-    result.args === EMPTY_VECTOR || return result.args # already cached
-    argtypes, vargs = get_argtypes(result.linfo)
-    result.args = argtypes
-    if vargs !== nothing
-        result.vargs = vargs
+# In theory, there could be a `cache` containing a matching `InferenceResult`
+# for the provided `linfo` and `given_argtypes`. The purpose of this function is
+# to return a valid value for `cache_lookup(linfo, argtypes, cache).argtypes`,
+# so that we can construct cache-correct `InferenceResult`s in the first place.
+function matching_cache_argtypes(linfo::MethodInstance, given_argtypes::Vector)
+    toplevel = !isa(linfo.def, Method)
+    @assert !toplevel
+    nargs::Int = linfo.def.nargs
+    @assert length(given_argtypes) >= (nargs - 1) # TODO do we need this assert
+    result_argtypes = Vector{Any}(undef, length(given_argtypes))
+    for i in 1:length(given_argtypes)
+        a = given_argtypes[i]
+        if !isa(a, PartialTuple) && !isa(a, Const)
+            a = widenconst(a)
+        end
+        result_argtypes[i] = a
     end
-    return argtypes
+    if linfo.def.isva
+        isva_result_argtypes = Vector{Any}(undef, nargs)
+        for i = 1:(nargs - 1)
+            isva_result_argtypes[i] = result_argtypes[i]
+        end
+        isva_result_argtypes[nargs] = tuple_tfunc(result_argtypes[(nargs - 1):end])
+        return isva_result_argtypes
+    end
+    return result_argtypes
```

Comment:
hmmm replacing this whole method body with a call to matching_cache_argtypes(linfo, nothing) at least allows it to build past the point that I was getting a BoundsError before (though of course probably results in a bunch of cache misses)